Cache GTFS metadata and expose utility attributes (breaking change)

renemarc opened the pull request:
## Description:
Current sensor updates run 7 additional SQLite database queries to populate attributes, on top of the bus schedule queries themselves. Double that if you have two sensors. That leads to a lot of slowdowns for everything else when using an SD card!

Considering that some data never changes (agency, routes...) and that others like departure times are good until invalidated, let's fetch such metadata at first then only when relevant changes do occur.

**Breaking Change:**
GTFS sensor attributes are now named using the standard snake_case format.

### Work performed:
- All metadata queries are now cached.
- Metadata queries are now all regrouped in the `update()` method.
- Attributes assembling is now done in ~~`device_state_attributes()` where it belongs.~~ in a utility method called from `update()`, for code clarity and since there is potential I/O from SQLAlchemy.
- As a bonus, many metadata entries with cryptic values have complementary entries added that provide easier to use data:
	- .\* Stop Drop Off Type: .\* Stop Drop Off Type **State** -> (string, unknown)
	- .\* Stop Pickup Type: .\* Stop Pickup Type **State** -> (string, unknown)
	- .\* Stop Timepoint: .\* Stop Timepoint **Exact** -> boolean
	- .\* Station Location Type: .\* Station Location Type **Name** -> string
	- .\* Wheelchair Boarding: .\* Wheelchair Boarding **Available** -> (boolean, unknown)
	- Route Type: Route Type **Name** (string)
	- Trip Bikes Allowed: Trip Bikes Allowed **State** -> (boolean, unknown)
	- Trip Wheelchair Access: Trip Wheelchair Access **Available** -> (boolean, unknown)
- Attribute names are now using snake_case.
- Added type hints.

**Related issue (if applicable):** fixes #21222

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

renemarc committed:
Cache GTFS attributes

* Cache all metadata queries

* Convert attribute names to snake_case (breaking change)

* Separate attributes update into its own method

* Expose human-readable equivalents for some cryptic attributes

* Add typing

* Revise update() method

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -122,41 +189,37 @@ def get_next_departure(sched, start_station_id, end_station_id, offset):
     }
 
-    destination_stop_time_dict = {
+    destination_stop_time = {
```
Opportunistic name simplification

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -122,41 +189,37 @@ def get_next_departure(sched, start_station_id, end_station_id, offset):
-    route = sched.routes_by_id(item['route_id'])[0]
-
-    origin_stop_time_dict = {
+    origin_stop_time = {
```
Opportunistic name simplification

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,193 @@ def unit_of_measurement(self):
     @property
     def icon(self):
         """Icon to use in the frontend, if any."""
+        self._icon = ICONS.get(self._route_type, ICON)
```
Moved here from update().

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,193 @@ def unit_of_measurement(self):
     def update(self):
         """Get the latest data from GTFS and update the states."""
         with self.lock:
+            # Fetch valid stop information once
```
Stops lookups moved here from `get_next_departure()`, and performed only once.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,193 @@ def unit_of_measurement(self):
-            trip = self._departure['trip']
+            route_id = self._departure['route_id']
+            trip_id = self._departure['trip_id']
+
```
Metadata objects all loaded here for later processing as attributes.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
     @property
     def icon(self):
         """Icon to use in the frontend, if any."""
+        self._icon = ICONS.get(self._route_type, ICON)
```
Moved here from `update()`.

balloob replied:
properties should not have side effects, so keep this inside `update()`

renemarc replied:
OK, reverted!

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
     def update(self):
         """Get the latest data from GTFS and update the states."""
         with self.lock:
+            # Fetch valid stop information once
```
Stops lookups moved here from `get_next_departure()`, and performed only once.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
-            agency = self._departure['agency']
-            route = self._departure['route']
-            trip = self._departure['trip']
+            # Fetch metadata details once, unless updated
```
Metadata objects all loaded from database here for later processing as attributes in `update_attributes()` below. Loading is done only once unless references have changed.

balloob commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
     def device_state_attributes(self):
         """Return the state attributes."""
-        return self._attributes
+        return {k: self._attributes[k] for k in sorted(self._attributes)}
```
Revert this please. Also, this won't work in Python 3.5 as dictionaries are only ordered starting 3.6

balloob replied:
Backend should not do anything in the code to help the frontend. Frontend should do such things.

renemarc replied:
Acknowledged, reverted!

balloob commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
+            if not self._origin:
+                stops = self._pygtfs.stops_by_id(self.origin)
+                if not stops:
+                    raise ValueError(
```
What's the use of raising a ValueError if no one catches it? You should catch it and update the state accordingly. I suggest that if an error is raised, you return `False` from the `available` property.

renemarc replied:
The goal was to error the component at start-up and prevent further attempts at fetching a non-existant database entry. However I can indeed exit this method gracefully; this way should the database record gets fixed between updates then the method will eventually provide a useful value.

renemarc replied:
Fixed: now returning instead of throwing an exception.

balloob commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
+            if not self._destination:
+                stops = self._pygtfs.stops_by_id(self.destination)
+                if not stops:
+                    raise ValueError("Destination stop ID {} not found".format(
```
idem

renemarc replied:
Done!

balloob commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
+
+    def update_attributes(self):
+        """Update state attributes."""
+        def dict_for_table(resource) -> dict:
```
Move these out of this function definition

renemarc replied:
Moved out and made `dict_for_table()` static.

balloob commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
+            return dict((col, getattr(resource, col))
+                        for col in resource.__table__.columns.keys())
+
+        def append_keys(resource, prefix=None) -> None:
```
Move out of this function definition. Also, do not use pretty keys but instead use key lowercase and underscores. The backend should not be aware of the frontend.

renemarc replied:
Will do! However this will introduce a breaking change as pretty keys have been in use with this component since its introduction.

renemarc replied:
Fixed! I'll add a comment regarding the breaking change in the PR description.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -4,30 +4,61 @@
-import os
-import logging
 import datetime
+import logging
```
isorted.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -213,104 +278,255 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     @property
-    def name(self):
+    def name(self) -> str:
```
Opportunistic type hinting (and all methods below). Further type hinting to be performed in upcoming PRs.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -52,11 +123,9 @@
 
 
-def get_next_departure(sched, start_station_id, end_station_id, offset):
+def get_next_departure(schedule: Any, start_station_id: Any,
```
Opportunistic type hinting (and all functions/methods below too).

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -208,109 +277,269 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            # Assign attributes, icon and name
+            self.update_attributes()
+
+            if self._route_type is not None:
```
Added defensive check.

renemarc commented on homeassistant/components/sensor/gtfs.py:
```diff
@@ -208,109 +277,269 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            self.remove_keys(prefix)
+
+    @staticmethod
+    def dict_for_table(resource: Any) -> dict:
```
Type-hinted as `Any` because SQLAlchemy declarative bases are voodoo. 😨

robbiet480 commented:
@renemarc Same as the other PR, just fix the conflicts and its good to merge

renemarc commented:
> @renemarc Same as the other PR, just fix the conflicts and its good to merge

Thanks again @robbiet480! I've rebased the branch (and added some logger lines), but do expect further small conflicts once the other PR is merged. I'll keep this PR updated as they appear.

robbiet480 commented:
@renemarc This one is ready to rebase

renemarc commented:
@robbiet480: I've rebased the branch and thanks to the recent merges I was able to make the following notable improvements from the previous state:
- Refactored the DB result column `origin_departure_date` to `origin_depart_date` for consistency.
- Removed the early exit from `update()` in favour of a flattened structure and reduced code duplication.
- Eliminated the now redundant private property `self._route_type`.
- Transformed the remaining in-line attribute keys into constants.
- Merged the sensor naming code into a single block (a later PR will address the entity_id juggle issue #18008).

Happy reviewing! 😃

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     add_entities([
         GTFSDepartureSensor(gtfs, name, origin, destination, offset,
                             include_tomorrow)])
+    return True
```
Nothing is checking this return value.

renemarc replied:
Fixed!

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -237,49 +313,47 @@ def get_next_departure(sched, start_station_id, end_station_id, offset,
-def setup_platform(hass, config, add_entities, discovery_info=None):
+def setup_platform(hass: HomeAssistantType, config: ConfigType,
+                   add_entities: Callable[[list], None],
+                   discovery_info: Optional[dict] = None) -> bool:
```
Return value is wrong. This function shouldn't return a boolean. We shouldn't return anything explicitly.

renemarc replied:
I could fix to return `None` maybe, like [group/light.py](https://github.com/home-assistant/home-assistant/blob/e70931da67381664a1bfb1bae10a914f7b9780b6/homeassistant/components/group/light.py#L44) and [switch/light.py](https://github.com/home-assistant/home-assistant/blob/e70931da67381664a1bfb1bae10a914f7b9780b6/homeassistant/components/switch/light.py#L31) do?

MartinHjelmare replied:
Yes, it should be `-> None:`.

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -237,49 +313,47 @@ def get_next_departure(sched, start_station_id, end_station_id, offset,
     """Set up the GTFS sensor."""
     gtfs_dir = hass.config.path(DEFAULT_PATH)
-    data = config.get(CONF_DATA)
+    data = str(config.get(CONF_DATA))
```
No need to copy to string. Use `dict[key]` for required config keys and keys with default config schema values.

renemarc replied:
This type casting addition was to prevent **Optional[Any]** typing warnings when feeding the data variable to `os.path.join` which accepts **Union[str, _PathLike[str]]**.

I can add a `# type: str` comment instead, considering that `cv.string` will validate the type for us beforehand.

renemarc replied:
...or indeed simply use `config[CONF_DATA]` like you suggest, must less of a typing rabbit hole and makes for cleaner code. 😃

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    def state(self) -> str:
         """Return the state of the sensor."""
+        if self._state is None:
+            return STATE_UNKNOWN
```
Return `None` to represent unknown state.

renemarc replied:
Returning `None` however introduces an inconsistency regarding Entity's state property return type:
https://github.com/home-assistant/home-assistant/blob/e70931da67381664a1bfb1bae10a914f7b9780b6/homeassistant/helpers/entity.py#L109-L113

Sort of modifying Entity's state property to return **Optional[str]**, I could instead define GTFS's state property as Optional[str], but also invalidates the interfacing somewhat. Hence my returning of **STATE_UNKNOWN**.

I'm curious, which cleanest path should we take?

MartinHjelmare replied:
We should `return None` and let the base entity class handle the unknown state conversion.

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            else:
+                trip_id = self._departure['trip_id']
+                if not self._trip or self._trip.trip_id != trip_id:
+                    _LOGGER.info("Fetching trip details for %s", trip_id)
```
Debug level max.

renemarc replied:
ok!

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+
+                route_id = self._departure['route_id']
+                if not self._route or self._route.route_id != route_id:
+                    _LOGGER.info("Fetching route details for %s", route_id)
```
Same as above.

renemarc replied:
ok!

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            # Fetch agency details exactly once
+            if self._agency is None and self._route:
+                try:
+                    _LOGGER.info("Fetching agency details for %s",
```
Same as above.

MartinHjelmare replied:
Why wrap the logging in the try... except?

renemarc replied:
Some agencies [do not set their agency_id in the `routes` table](https://transitfeeds.com/p/edmonton-transit-system/27/latest/file/routes.txt), so all values for the `agency_id` SQLite column end up as `None`. Prior to this PR [the update() would simply fail](https://github.com/home-assistant/home-assistant/blob/eb3355b05dd0edac38ee0e33a6d007b61c338fe9/homeassistant/components/sensor/gtfs.py#L169) if one didn't manually fix this missing relationship using something like:
```sql
UPDATE routes
SET agency_id = (
   SELECT agency_id
   FROM agency
   WHERE feed_id = 1
)
WHERE agency_id = 'None';
```

Now at least the sensor doesn't fail for such a non-critical metadata query, but a warning is issued about this missing relationship. A further PR could implement a pro-active fix at the database creation step.

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                    _LOGGER.warning(
+                        "Agency ID '%s' not found in agency table. You may "
+                        "want to update the agency database table to fix this "
+                        "missing reference.", self._route.agency_id)
```
We don't end logging messages with period.

renemarc replied:
Will fix in next PR!

MartinHjelmare commented on homeassistant/components/gtfs/sensor.py:
```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+
+            if self._departure[ATTR_FIRST] is not None:
+                self._attributes[ATTR_FIRST] = self._departure['first']
+            elif ATTR_FIRST in self._attributes.keys():
```
Drop `.keys()`. It's not needed.

renemarc replied:
ok!

MartinHjelmare reviewed (commented):
Please open a new PR where we can address the comments.

For the future, please don't mix fixes, typing, features or other types of changes in the same PR. One type of change per PR and keep it as small as possible. Mixed changes like this is hard to review.

renemarc commented:
> Please open a new PR where we can address the comments.
> 
> For the future, please don't mix fixes, typing, features or other types of changes in the same PR. One type of change per PR and keep it as small as possible. Mixed changes like this is hard to review.

Will do! Indeed, this was a rather large-ish PR. I'll try to keep others in check.
